### ISSUE TITLE
Improve Static Analysis

### ISSUE THREAD

**Issue by sraboy** [angr/angr#575]
I was wondering if there is any interest in expanding on angr's use for static analysis? I'm proposing something like a new `project.analyses.static` class. 

I wrote up a `StaticAnalysis` class (and some supporting ASCII table generators) that, for now, takes a `cle.backends.Backend` instance:

```
In [723]: ps = angr.Project('./somelib.o')
     ...: ss = StaticAnalysis(ps.loader.main_object)
     ...: ss.ShowTable('binprops')
     ...: ss.ShowTable('sections')
     ...: ss.ShowTable('imports')
     ...: ss.ShowTable('relocs')
     ...: 
WARNING | 2017-08-10 16:37:02,290 | cle.elf | no segments identified in PT_LOAD
WARNING | 2017-08-10 16:37:02,372 | cle.loader | The main binary is a position-independent executable. It is being loaded with a base address of 0x400000.
DEBUG   | 2017-08-10 16:37:02,374 | angr.static | Got props [['Filename', 'somelib.o'], ['Arch', 'Arch ARMEL (LE)'], ['Bitness', '32'], ['OS', 'unix'], ['PIC/PIE', 'True'], ['Linked Base', '0x000000'], ['Mapped Base', '0x400000'], ['Linking', 'static']]
-----------------------------------
 |       Binary Properties       | 
-----------------------------------
 |  Property   |      Value      | 
-----------------------------------
 | Arch        | Arch ARMEL (LE) | 
 | Bitness     | 32              | 
 | Filename    | somelib.o       | 
 | Linked Base | 0x000000        | 
 | Linking     | static          | 
 | Mapped Base | 0x400000        | 
 | OS          | unix            | 
 | PIC/PIE     | True            | 
-----------------------------------
-------------------------------------------------------------------------------------------------------------------------------
 |                                                         Sections                                                          | 
-------------------------------------------------------------------------------------------------------------------------------
 |      Name       |  Offset  |  Virt Addr  |  Size  |  Mem Size  |  Align  |  Entry Size  |   Flags    |        Type        | 
-------------------------------------------------------------------------------------------------------------------------------
 |                 | 0x000000 |  0x400000   |   0    |     0      |    0    |      0       |    0x0     | SHT_NULL           | 
 | .ARM.attributes | 0x0001f8 |  0x400000   |   49   |     49     |    1    |      0       |    0x0     | SHT_ARM_ATTRIBUTES | 
 |            .bss | 0x000178 |  0x400144   |   0    |     0      |    1    |      0       |    0x3     | SHT_NOBITS         | 
 |        .comment | 0x0001d8 |  0x400000   |   32   |     32     |    1    |      1       |    0x30    | SHT_PROGBITS       | 
 |           .data | 0x000178 |  0x400144   |   0    |     0      |    1    |      0       |    0x3     | SHT_PROGBITS       | 
 | .note.GNU-stack | 0x0001f8 |  0x400000   |   0    |     0      |    1    |      0       |    0x0     | SHT_PROGBITS       | 
 |       .rel.text | 0x0003a4 |  0x400000   |  112   |    112     |    4    |      8       |    0x40    | SHT_REL            | 
 |         .rodata | 0x000178 |  0x400144   |   96   |     96     |    4    |      0       |    0x2     | SHT_PROGBITS       | 
 |       .shstrtab | 0x000229 |  0x400000   |   97   |     97     |    1    |      0       |    0x0     | SHT_STRTAB         | 
 |         .strtab | 0x00037c |  0x400000   |   38   |     38     |    1    |      0       |    0x0     | SHT_STRTAB         | 
 |         .symtab | 0x00028c |  0x400000   |  240   |    240     |    4    |      16      |    0x0     | SHT_SYMTAB         | 
 |           .text | 0x000034 |  0x400000   |  324   |    324     |    4    |      0       |    0x6     | SHT_PROGBITS       | 
-------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------
 |                                     Imports                                      | 
--------------------------------------------------------------------------------------
 |  Symbol Name  |  Dest Addr  |  Linked Addr  |  Rebased Addr  |  Relocation Type  | 
--------------------------------------------------------------------------------------
 |          puts |  0x00008c   |   0x000000    |    0x400000    | R_ARM_CALL        | 
 |        printf |  0x000104   |   0x000000    |    0x400000    | R_ARM_CALL        | 
--------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------
 |                                   Relocations                                    | 
--------------------------------------------------------------------------------------
 |  Symbol Name  |  Dest Addr  |  Linked Addr  |  Rebased Addr  |  Relocation Type  | 
--------------------------------------------------------------------------------------
 |               |  0x000128   |   0x000144    |    0x400144    | R_ARM_REL32       | 
 |               |  0x00012c   |   0x000144    |    0x400144    | R_ARM_REL32       | 
 |               |  0x000130   |   0x000144    |    0x400144    | R_ARM_REL32       | 
 |               |  0x000134   |   0x000144    |    0x400144    | R_ARM_REL32       | 
 |               |  0x000138   |   0x000144    |    0x400144    | R_ARM_REL32       | 
 |               |  0x00013c   |   0x000144    |    0x400144    | R_ARM_REL32       | 
 |               |  0x000140   |   0x000144    |    0x400144    | R_ARM_REL32       | 
 |        printf |  0x000048   |   0x000000    |    0x400000    | R_ARM_CALL        | 
 |        printf |  0x000064   |   0x000000    |    0x400000    | R_ARM_CALL        | 
 |        printf |  0x0000c8   |   0x000000    |    0x400000    | R_ARM_CALL        | 
 |        printf |  0x0000f0   |   0x000000    |    0x400000    | R_ARM_CALL        | 
 |        printf |  0x000104   |   0x000000    |    0x400000    | R_ARM_CALL        | 
 |          puts |  0x000078   |   0x000000    |    0x400000    | R_ARM_CALL        | 
 |          puts |  0x00008c   |   0x000000    |    0x400000    | R_ARM_CALL        | 
--------------------------------------------------------------------------------------
```

Ideally, it would be nice to be able to easily spit out more things, like: a function's disassembly, IR, callers, callees; meta properties (if it takes `angr.Project` instead) like constraints or hooks; missing symbol refs (like a to-do list before executing).

I'm not sure how much time I'll have to maintain/improve upon it in the future though. Maybe some sort of plugin architecture?

**Comment by rhelmot** [comment]
Can you explain what your proposed static analysis class would actually do besides have built-in pretty printing for results? We designed the base analysis class to be the place where you build static analyses, so honestly any changes you wanted to make could probably just be added to that. The sample analyses you show are honestly exactly like the examples we usually write to show off how to build analyses using the Analysis class.

If you're talking about the usability of the knowledge base, that's a very different point - our vision for the knowledge base is to have it be more of a database in the sense that it contains less "cached" data and more of a ground truth for what is _known_ about the program as a result of analysis, and then build a series of "views" on top of this database that provide all the supplementary stuff that you currently get from the knowledge base, like the graph of a function (the kb itself should only store the list of functions, the starts of basic blocks, and the targets of indirect jumps).

**Comment by sraboy** [comment]
Short answer: I'm looking to replace readelf, objdump, nm, dumpbin and IDA scripts.  

Long answer: Part of what I'm doing for work right now is evaluating whether or not angr (and its supporting tools) can replace my other tools, and to what extent. In that sense, I know I'm probably getting a bit out of scope with your focus for CGC, hence the mention of plugin architecture.

Since mitigating path explosion on non-trivial binaries requires so much static analysis work, I need to evaluate angr's ability to do that (i.e., without resorting to a $10K tool like Hex-Rays Decompiler). I'm using my class to essentially replace all those tools above by generating tabulated data I can manipulate into useful information with Python. For now, I pretty-print it since there's still too much that requires some human reasoning but, ideally, I can set some filters and automatically spit out a list of addresses to `find` or `avoid` with `simgr` and a list of memory locations I probably need claripy to resolve.

Right now, I need meticulous static RE to answer a question like, which of these pass what data to which libraries and where does that data come from? Like IDA, Angr and CLE have most of this information, it's just not terribly easy to get at without a CFG in a GUI. 

My [CLE PR](https://github.com/angr/cle/pull/79) helps with the first part of that question. Next is getting a list of memory locations in the caller that are used as parameters to the callee and figuring out what's in there, maybe with claripy and a `call_state` in the middle of the function using all those previously identified `avoid` addresses.

**Comment by ltfish** [comment]
I like your effort of pretty-printing the basic information of a binary, and that is indeed some nice functionalities that are currently missing in angr. IMO they are perfect for CLE. However, as far as I see them, those functionalities are only at the single-binary level, and there is not much analysis involved, so I would be reluctant to give them the awesome class name `StaticAnalysis`. If you make a PR for CLE, I will happily take it.

angr does a lot in terms of static analysis (although it's still far from sufficient). angr recovers CFG, some sort of data dependency information, and performs value-set analysis. We also have some [interesting static-analysis stuff](https://github.com/angr/angr/tree/wip/nightmaRE) that is in active development. All of those analysis routines have their own names. Creating an analysis class called `Static` or `StaticAnalysis` will confuses users.

> I'm looking to replace readelf, objdump, nm, dumpbin and IDA scripts.

Replacing `readelf`, `objdump`, `nm`, and `dumpbin` is a realistic goal. Replacing IDA scripts, if I understand it correctly, is equivalent to replacing IDA. That sounds *very* ambitious.

> Since mitigating path explosion on non-trivial binaries requires so much static analysis work, I need to evaluate angr's ability to do that (i.e., without resorting to a $10K tool like Hex-Rays Decompiler).

Are you aware of a project called [angr management](https://github.com/angr/angr-management/)? And are you aware of the `wip/nightmaRE` branch? They might be interesting to you.

> which of these pass what data to which libraries and where does that data come from? Like IDA, Angr and CLE have most of this information, it's just not terribly easy to get at without a CFG in a GUI.

You should really be consider put some development effort into angr management. It is still a piece of sh*t as of now, but (at least IMHO) it will soon be much closer to IDA than a console-based CFG visualization tool will be.

PS: if IDA's price is a concern to you, you might want to try [Binary Ninja](https://binary.ninja/). It is much cheaper, is in active development, and offers some really nice functionalities for both reversing (like medium-level IL) and interoperability (like a nice programming interface).

**Comment by sraboy** [comment]
I wasn't able to get angr-management running here but I haven't put a lot of effort into it so maybe I'll give that another try. I agree that only focusing on a single binary is most limiting; expanding that across multiple binaries would be a huge leg up over IDA. I'll take a look at nightmaRE to see what's already in there too.

As far as IDA goes, it's a great tool, especially with IDAPython. While IDA's internal analysis isn't something I could really duplicate or improve upon, I think a lot of the same information could be exposed, to the extent that angr/CLE have the fundamental analysis done. 

Some things I can do with IDA that could be duplicated by angr/CLE:

* Assign pseudo-symbols to functions and strings (allows renaming, bringing along the "interactive" from IDA.
* I've got an IDAPython script that I use to identify all jump tables, or just a jump table at some address, and spit out the addresses for each target location. I then just regex this into a Python list to make a nice 'avoid' list for `simgr` once I've identified the few cases I care about it.
* Identify symbol cross-references (probably the hardest thing in this list)
* Signature scanning based on [open-source FLIRT signatures](https://github.com/Maktm/FLIRTDB)
* Simplified disassembly (e.g., `p.get_symbol('somefunc').showdisasm(flavor='intel')`)

To do most of this work cross-platform, the API for ELF and PE backends needs to be better standardized. I'm refactoring PE and WinReloc now, so I can get standard information from ELF, MetaELF, PE and Relocation but I can tell everything was mostly written with ELF in mind so there's a lot that would need to be abstracted out to the base Backend, Section and Relocation classes.

Anyway, just my $0.02. I know yet another major refactor may not be something you guys want to take on at the moment so, for now, I'm keeping away from altering the base classes and just expanding/improving upon PE and WinReloc, to mirror the implementations from the ELF side. I've already added all the PE relocations to archinfo and new arch modules for relocations (named "peamd64", "pearm", etc)

Once I work out some kinks in the ASCII table printing, I'll put in a PR for CLE to at least get some review on the static analysis pieces and figure out exactly how to name it or expose the API. I'll do  separate PRs for the Windows stuff.

**Comment by ltfish** [comment]
> - Assign pseudo-symbols to functions and strings (allows renaming, bringing along the "interactive" from IDA.
> - I've got an IDAPython script that I use to identify all jump tables, or just a jump table at some address, and spit out the addresses for each target location. I then just regex this into a Python list to make a nice 'avoid' list for simgr once I've identified the few cases I care about it.
> - Identify symbol cross-references (probably the hardest thing in this list)
> - Simplified disassembly (e.g., p.get_symbol('somefunc').showdisasm(flavor='intel'))

All of these are already doable in angr. What is missing is just a nice interface and the documentation!

> - Signature scanning based on open-source FLIRT signatures

We don't have any implementation for FLIRT. It is a very nice thing to have!

>  I know yet another major refactor may not be something you guys want to take on at the moment so

I don't think adding new interfaces is a major refactor. In fact, since there are already some compatibility-breaking changes with the current refactor, we think this is the right time to make other compatibility-braking changes, too. We will be more reluctant to do so once the current refactor is completed.

> Once I work out some kinks in the ASCII table printing, I'll put in a PR for CLE to at least get some review on the static analysis pieces and figure out exactly how to name it or expose the API.

Sound great!

> I'll do separate PRs for the Windows stuff.

A suggestion: maybe you want to hop on to our Slack channel. We don't (yet) have the manpower to post updates of our development plans and progresses to GitHub, but we *do* talk about what is going on and what will be worked on in the Slack channel. For example, @rhelmot just merged some CLE refactor as well as some improved Windows executables support to master. It wasn't announced anywhere but in the Slack channel (since there was no major interface change anyway). You might want to have a look at his work before sending your PRs :)

In the end, thanks for your interest in angr! Replacing IDA is absolutely more difficult than building Rome, but we will be there one day.

**Comment by ltfish** [comment]
> I wasn't able to get angr-management running here

Let me know if there is any trouble getting it running. It should run out of the box (as long as you follow the README), otherwise it's my fault!

**Comment by rhelmot** [comment]
For your "interactive renaming" request, we have project.kb.labels for this, but it's not very well integrated.

**Comment by rhelmot** [comment]
For your disassembly needs we have the "Disassembly" analysis that I started writing to basically do what you want as well as be a modularized customizable pretty printer that could be used to power angr-management and command-line disassembly + graph views, but I got stuck and tossed it over to fish, who has taken it more in the direction of angr-management. You should look at that!

**Comment by ltfish** [comment]
> For your "interactive renaming" request, we have project.kb.labels for this, but it's not very well integrated.

Press `N` key in angr management! That's totally interactive.

**Comment by ltfish** [comment]
> but I got stuck and tossed it over to fish, who has taken it more in the direction of angr-management.

Hey, GUI is the future.

**Comment by sraboy** [comment]
I'm definitely going to explore kb and analyses more and I'll be sure to check out the latest from master to see the work done on Windows.

I'm actually terrible at UX design but I may tinker with angr-management, at least on the back end.

Just hopped on Slack so I'll keep an eye out!

**Comment by sraboy** [comment]
Rather than re-inventing the wheel for pretty-printing, would you guys be open to adding [terminaltables](https://pypi.python.org/pypi/terminaltables).

**Comment by ltfish** [comment]
> Rather than re-inventing the wheel for pretty-printing, would you guys be open to adding terminaltables?

Feel free to add it as a dependency in your PR!

**Comment by ltfish** [comment]
> I'm actually terrible at UX design but I may tinker with angr-management, at least on the back end.

I got a brilliant designer (@tiffanyb) to do the UX work, so everyone else can just focus on the backend stuff ;)

**Comment by sraboy** [comment]
I've got FLIRT signature recognition worked out by parsing SIG files with [nampa](https://github.com/kenoph/nampa). I need to do a PR to nampa to support v10 SIG files but, beyond that, it appears to work well. I still have to spend some time on "referenced functions" but need to get a few test cases working first.

I have three methods, each of which work with either a single SIG file or a directory of them:
1. Scan addrs in `[f for f in main_object.symbols_by_addr.values() if f.is_function is True]`
2. Scan addrs by `[f for f in kb.functions.values() if f.is_syscall is False]`
3. Scan addrs by `[f.addr for f in cfg.nodes() if f.name is None]`

Should no. 2 also exclude `f.is_simprocedure`? It appears that's only set true if it's hooked so I figured it should still be scanned.

Should this just be a separate module in Analyses? What should the interface be like? I'm thinking something like this:
`proj.analyses.sigscan(method='flirt', rename=True, scanopts={'opt1':'thing'...})`

Using `sigscan` is just to leave room for YARA, generic pattern matching, etc. `rename` specifies whether to rename the `sub_*` matches in the `kb` or not. In any case, a list of `Match` objects is returned; the objects allow the user to see what matched and how: 
- Matches a symbol by name or addr, but not both (FLIRT says FuncX is at `0xcafebabe` but a symbol for FuncX says it's at `0xdeadbeef`)
- Matches a symbol name _and_ addr
- Matches a function in `kb.functions` with name X at addr Y (may or may not match a symbol)
- Matches a non-function at some addr

I'm unsure if that last case should result in creating a label, creating a `function` in the `kb` or just warning the user.

**Comment by ltfish** [comment]
Awesome!

> Should no. 2 also exclude `f.is_simprocedure`? It appears that's only set true if it's hooked so I figured it should still be scanned.

It should exclude `f` if `f.is_simprocedure is True`. All angr-specific library functions are all `SimProcedure`s.

> 3. Scan addrs by [f.addr for f in cfg.nodes() if f.name is None]

I don't think this is needed. Maybe we can make this step optional?

> Should this just be a separate module in Analyses?

Yes. I hope there is a more impressive name than `sigscan`.

**Comment by sraboy** [comment]
Okay, I'll drop the simprocedures from the scan. I should have been clearer on the scan methods: each is just an option and the last is more of a last resort but could help with odd binaries where a function isn't recognized by CFG for some reason.

I've been trying to think of a good name for the class... Maybe I'll ping Slack.

**Comment by sraboy** [comment]
Just wanted to close the loop here since I don't have as much time to work on this now. This is my wish list and some partially-done work on the issue.

- [ ] Standardized backends for ELF/PE: _On hold_.  
     - [ ] Refactor PE to more closely match ELF's implementation. Waiting on evaluation and merge of [CLE PR 80](https://github.com/angr/cle/pull/80).
     - See the to-do list in that issue for more
- [ ] FLIRT Signatures: _On hold_. Currently [works](https://github.com/sraboy/angr/tree/wip/flirt) with a [modified nampa](https://github.com/sraboy/nampa/tree/wip/reffuncs) but, for large binaries, it's incredibly slow. 
     - [x] Basic scanning: potential for lots of false-positives for complex binaries.
     - [ ] CFG Refinement: Needs #611. Current kludge is to re-generate the CFG for every function identifieid, which can be painfully slow.
     - [ ] REF Funcs: Will need changes to nampa (See [nampa issue #4](https://github.com/kenoph/nampa/issues/4))
     - [ ] Startup signature support: Would also need changes to nampa. 
- [ ] `static` module: _In Progress_. Either in `analyses` or as a kb plugin, this would have some common APIs to access information from `elftools` and `pefile` through a common `backend.reader` attribute. It will provide simple lists/dictionaries of information that can be ingested/processed elsewhere, like angr-management, as well as pretty-printed to the terminal or a text file.
     - [x] Submission to VirusTotal and retrieval of results
     - [x] `strings` like output
     - [ ] Simple dict for imports/deps: `{'func': 'somelib'}` or `{'somelib': ['func1', 'func2'] }`
     - [ ] `base_report`: Simple dict of file/binary attributes like name, size, type, compile date/time, compiler identification, deps, hashes, function count.
     - [ ] Pretty printing through [terminaltables](https://pypi.python.org/pypi/terminaltables)
     - [ ] Arbitrary signature scanning: a la [Detect It Easy](https://github.com/horsicq/Detect-It-Easy)
     - [ ] Compiler/linker detection: FLIRT startup signatures or some other method
- [ ] Simple function disassembly: Something like `kb.functions[addr].disasm(syntax='intel')`

As I have time, for `static`, I'm mostly just running through Practical Malware Analysis' labs (binaries [here](https://github.com/mikesiko/PracticalMalwareAnalysis-Labs)) and seeing how I can answer its questions solely through an angr project and writing functions to do it for me.

**Comment by otelo29200** [comment]
>> -  Identify symbol cross-references (probably the hardest thing in this list)
>>
> All of these are already doable in angr. What is missing is just a nice interface and the documentation!

Hey ! How would you do this through angr API ? Or should I do it with a function calling capstone, something like this ?

**Comment by ltfish** [comment]
> Hey ! How would you do this through angr API ? Or should I do it with a function calling capstone, something like this ?

There isn't a nice API to do the querying. But the information is already stored in a `CFGFast` instance, if you enable the `collect_data_references` option when building the CFG.
